gcc/rust/gcc-rust/src/gcc_api.rs

```diff
@@ -2050,51 +81,70 @@ impl Tree {
     }
 
     pub fn new_function_type(return_type: Tree, arg_types: &[Tree]) -> Self {
-        unsafe { _build_function_type_array(return_type, arg_types.len(), arg_types.as_ptr()) }
+        unsafe {
+            Tree(build_function_type_array(
+                return_type.0,
+                arg_types.len() as i32,
+                arg_types.as_ptr() as *mut tree,
+            ))
+        }
+    }
+
+    pub fn new0(code: TreeCode, type_: Tree) -> Self {
+        unsafe { Tree(build0(code, type_.0)) }
     }
 
     pub fn new1(code: TreeCode, type_: Tree, arg0: Tree) -> Self {
-        unsafe { _build1(code, type_, arg0) }
+        unsafe { Tree(build1(code, type_.0, arg0.0)) }
     }
 
     pub fn new2(code: TreeCode, type_: Tree, arg0: Tree, arg1: Tree) -> Self {
-        unsafe { _build2(code, type_, arg0, arg1) }
+        unsafe { Tree(build2(code, type_.0, arg0.0, arg1.0)) }
     }
 
-    fn new3(code: TreeCode, type_: Tree, arg0: Tree, arg1: Tree, arg2: Tree) -> Self {
-        unsafe { _build3(code, type_, arg0, arg1, arg2) }
+    pub fn new3(code: TreeCode, type_: Tree, arg0: Tree, arg1: Tree, arg2: Tree) -> Self {
+        unsafe { Tree(build3(code, type_.0, arg0.0, arg1.0, arg2.0)) }
+    }
+
+    pub fn new4(
+        code: TreeCode,
+        type_: Tree,
+        arg0: Tree,
+        arg1: Tree,
+        arg2: Tree,
+        arg3: Tree,
+    ) -> Self {
+        unsafe { Tree(build4(code, type_.0, arg0.0, arg1.0, arg2.0, arg3.0)) }
     }
 
     pub fn new_translation_unit_decl(name: Tree) -> Self {
-        unsafe { _build_translation_unit_decl(name) }
+        unsafe { Tree(build_translation_unit_decl(name.0)) }
     }
 
     pub fn new_init_expr(var: Tree, value: Tree) -> Self {
         Self::new2(TreeCode::InitExpr, TreeIndex::VoidType.into(), var, value)
     }
 
-    pub fn new_int_constant<T: Into<Tree>>(int_type: T, value: i64) -> Self {
-        unsafe { build_int_constant(int_type.into(), value) }
+    pub fn new_int_constant<T: Into<Tree>>(int_type: T, value: u64) -> Self {
```

Comment:
Why unsigned? The c++ code has poly_int64 and
```c++
typedef poly_int<NUM_POLY_INT_COEFFS, HOST_WIDE_INT> poly_int64;
typedef poly_int<NUM_POLY_INT_COEFFS, unsigned HOST_WIDE_INT> poly_uint64;
```
so I'm pretty sure that poly_int64 is supposed to be signed, which is good because this function really does need to support negative values.

btw, what does cbindgen do with the C++ type here? Does it call the constructor for us?

Reply:
I trusted bindgen for the unsigned but after looking at the doc poly_int seems more subtle than that.
It's value depend of the architecture, on most of them NUM_POLY_INT_COEFFS is 1 so poly_int64 is equivalent to HOST_WIDE_INT (which can represent both signed and unsigned number, so probably bindgen got confused with an internal representation somewhere).
However, for Aarch64 NUM_POLY_INT_COEFFS is 2 so we need specific code to handle that.
The value of NUM_POLY_INT_COEFFS is defined in a compile time generated header, (<build_dir>/gcc/insn-modes.h), so we will probably need something to parse it and pass a custom cfg attr to rustc.

Reply:
Added a build time panic for architecture with NUM_POLY_INT_COEFFS != 1 and changed poly_int64 to be a signed number.